asteroid/dsp/overlap_add.py

```diff
@@ -0,0 +1,147 @@
+        nnet,
+        n_src,
+        window_size,
+        hop_size=None,
+        window="hanning",
+        reorder_chunks=True,
+        enable_grad=False,
+    ):
+        super().__init__()
+        assert window_size % 2 == 0, "Window size must be even"
+
+        self.nnet = nnet
+        self.window_size = window_size
+        self.hop_size = hop_size if hop_size is not None else window_size // 2
+        self.n_src = n_src
+
+        if window:
+            window = get_window(window, self.window_size).astype("float32")
+            window = torch.from_numpy(window)
+            self.use_window = True
+        else:
+            self.use_window = False
+
+        self.register_buffer("window", window)
+        self.reorder_chunks = reorder_chunks
+        self.enable_grad = enable_grad
+
+    def ola_forward(self, x):
+        """Heart of the class: segment signal, apply func, combine with OLA."""
+
+        assert len(x.shape) == 3
+
+        batch, channels, n_frames = x.size()
+        # Overlap and add:
+        # [batch, chans, n_frames] -> [batch, chans, win_size, n_chunks]
+        folded = torch.nn.functional.unfold(
+            x.unsqueeze(-1),
+            kernel_size=(self.window_size, 1),
+            padding=(self.window_size, 0),
+            stride=(self.hop_size, 1),
+        )
+
+        out = []
+        n_chunks = folded.shape[-1]
+        for f in range(n_chunks):  # for loop to spare memory
+            tmp = self.nnet(folded[..., f])
+            # user must handle multichannel by reshaping to batch
+            assert len(tmp.size()) == 3, "nnet should return (batch, n_src, time)"
+            assert tmp.shape[1] == self.n_src, "nnet should return (batch, n_src, time)"
+            tmp = tmp.reshape(batch * self.n_src, -1)
+
+            if f != 0 and self.reorder_chunks:
+                # we determine best perm based on xcorr with previous sources
+                tmp = _reorder_sources(tmp, out[-1], self.n_src, self.window_size, self.hop_size)
+
+            if self.use_window:
+                tmp = tmp * self.window
+            else:
+                tmp = tmp / (self.window_size / self.hop_size)
+            out.append(tmp)
+
+        out = torch.stack(out).reshape(n_chunks, batch * self.n_src, self.window_size)
+        out = out.permute(1, 2, 0)
+
+        out = torch.nn.functional.fold(
+            out,
+            (n_frames, 1),
+            kernel_size=(self.window_size, 1),
+            padding=(self.window_size, 0),
+            stride=(self.hop_size, 1),
+        )
+        return out.squeeze(-1).reshape(batch, self.n_src, -1)
+
+    def forward(self, x):
+        """ Forward module: segment signal, apply func, combine with OLA.
+
+        Args:
+            x (:class:`torch.Tensor`): waveform signal of shape (batch, 1, time).
+
+        Returns:
+            :class:`torch.Tensor`: The output of the lambda OLA.
+        """
+        # Here we can do the reshaping
+        if self.enable_grad:
```

Comment:
I think you could use ` with torch.autograd.set_grad_enabled(self.enable_grad): ...` here

Reply:
cool I will switch to that I did not knew